### ISSUE TITLE
New Angular Router folder guidance

### ISSUE THREAD

**Issue by robarchibald** [johnpapa/angular-styleguide#386]
What recommendations do you have for using the new Angular Router?  By default it's wanting to put all components into a components/<componentName> folder for routing.  While that's closer to the folder structure you're using than the original "sock drawer" method of organization (http://cliffmeyers.com/blog/2013/4/21/code-organization-angularjs-javascript), it's not as nice as what you're doing with this style guide.  Do you recommend conforming to the new Angular Router structure?

**Comment by johnpapa** [comment]
This is something we'll have to keep looking at as the new router forces some things on us that are different and less flexible in structure. But we gain a lot with its features as it in itself is more flexible.

I don't have a good answer for you right now other than we (the community) have to give a few options a shot and then work it into the style guide, just like all of the other features in this guide.

Perhaps we keep this thread open to keep that discussion going.

**Comment by wardbell** [comment]
Remember that the conventions at configurable; we may discover that its worth changing the default config to match our preferences even at the cost of departing from the default.

There are many other, more significant aspects of its usage that we need to explore.

As John says, we need experience before we recommend.

**Comment by robarc** [comment]
Thanks Ward and John for the comments.  I actually went and looked again at how the new router is configured, and it doesn't look like it would be possible to *automatically* route the way that the Angular Style Guide suggests.  The style guide organization is by high-level "chunks" that go together, not by component.  The new router uses "convention over configuration" while the Angular Style Guide is implemented by configuring each route.  We could override the $componentLoaderProvider in the new router to configure every component to a new route in code, but I don't think that's needed.  As an example, here's the implementation in `$componentLoaderProvider` for the default location for html files: 

    var componentToTemplate = function componentToTemplateDefault(name) {
        var dashName = dashCase(name);
        return './components/' + dashName + '/' + dashName + '.html';
    };

For a real-world example, let's pretend we're creating an image gallery.  Perhaps I would create a "galleries" component to handle the list view and a "galleryDetail" component to handle the detail view of a single gallery.  With the guidance of the style guide, I'd probably group them together under a single "app/gallery" directory which makes sense as a human.  The default behavior of the new angular router puts them in separate directories.   If we wanted to get them unified, I think it would be hard to create a single, simple function to cover this case (We'd have to deal with plurals and singular names, not to mention all the other possibilities).  

Example | Angular Style Guide | New Angular Router
------------ | ------------ | -------------
Simple Galleries component | app/gallery | app/components/galleries
Simple Gallery component | app/gallery | app/components/gallery

And, to make things more complicated, the recommendation of the style guide is to break things into further folders once things get complicated (move beyond about 7 items in the folder).  This would yield a layout like this.  

Example | Angular Style Guide | New Angular Router
------------ | ------------ | -------------
Complex Galleries component | app/gallery/list | app/components/galleries
Complex Gallery component | app/gallery/detail | app/components/gallery

Hence, while the new angular router folder structure isn't as ideal for finding things as the human "configured" Angular Style Guide, I think the "convention" it will enforce is valuable.  And, if each component grows to the standard size of the style guide, each component should really have its own folder anyway (controller, controller spec, html, module, route, route spec).  That's 6 files and you recommend splitting into multiple folders at 7 anyway.  While it's not quite as tidy as the Angular Style Guide, it's an easy to remember and enforce convention.  And, if you name your components right, they will be close to one another in the folder tree so that like things will naturally be grouped together anyway (e.g. galleries and galleryDetail will be close to one another alphabetically)

So, my recommendation would be for the style guide to adopt the directory structure pattern recommended by the Angular 2 router.

**Comment by wardbell** [comment]
I'm not ready to concede. Or decide.  I only know that I don't like separating the JS from the template. I'll do it if I have to but I'd like to see what we can do first.

I'm kind of mystified by your pluralization example. How do you get this?
```
Simple Galleries component |	app/gallery | app/components/galleries
```

I DO appreciate your exploration though; please don't be put off by my reaction.

**Comment by johnpapa** [comment]
good research.

i'm not a fan of a component that imposes a directory structure on us. now convention can be overridden, but there are also ways in some routers (ones that Rob E wrote included) to override the conventions in one place, too. So that could be use to to balance making it easy to find code and for the router to work with it.

I'm not ready to dive deep here yet, but that's where I would start.

**Comment by robarc** [comment]
To answer Ward's question, the reason that I stated that the Simple "Galleries" component and "Gallery" component would both go into an app/gallery folder is based on recommendation [Y152](https://github.com/johnpapa/angular-styleguide#style-y152) from the Style Guide: "Create folders named for the feature they represent. When a folder grows to contain more than 7 files, start to consider creating a folder for them. Your threshold may be different, so adjust as needed."  The example in the style guide is almost the same as mine, except it was for sessions, not galleries:

    sessions/
        sessions.html
        sessions.controller.js
        sessions.routes.js
        session-detail.html
        session-detail.controller.js

Whether you choose to name the folder for its singular name (gallery) or plural name (galleries), both the singular and plural go together until it gets to the point where it is complicated enough to warrant adding additional folders.  

As for whether I was "put off by [your] reaction", I wasn't.  I'm just mystified that my stellar post didn't immediately cause all the stars to align to my view (ha, ha!).

**Comment by robarc** [comment]
So, I've been thinking about this more, and I thought I'd also look at the Hot Towel Yeoman generator as a second example.  It's a more fleshed out app than the example in the style guide.  Its directory structure is this for the admin component:

    admin/
        admin.controller.js
        admin.controller.spec.js
        admin.html
        admin.module.js
        admin.route.js
        admin.route.js

It looks to me that this number of files is going to be the minimum number of files needed for a new angular router app going forward, just like with the ui-router.  That's 6 files.  If that's the minimum for good style on a route-able component, then with the new router, we'd never put galleries and gallery components into the same folder (or sessions and session-detail in the style guide) since that'd mean we'd already be at 12 files at a minimum which exceeds the 7 file threshold before putting in separate folders.

If Hot Towel was built using the default routing from the new angular router, it'd look like the below:  Blocks, Core, Layout and Widgets are unchanged since they aren't components being routed to, but admin and dashboard both get moved under the components folder.  

    blocks/ 
        exception/
            4 files...
        logger/
             2 files...
        router/
             2 files...
    components/
        admin/
            admin.controller.js
            admin.controller.spec.js
            admin.html
            admin.module.js
            admin.route.js
            admin.route.js
        dashboard/
            dashboard.controller.js
            dashboard.controller.spec.js
            dashboard.html
            dashboard.module.js
            dashboard.route.js
            dashboard.route.spec.js
    core/
        7 files...
    layout/
        11 files...
    widgets/
        4 files...

Initially, I thought I liked this just fine, but the more I think about it, adding the "component" folder to the mix muddies the water significantly.  Now, half of the code is organized by technical type and half is organized by feature type.  How does a developer know where to put their code now?  For example, the layout folder has a shell and a sidebar that could be defined as a "component" in the technical sense of the word (a view + controller), but they are both being implemented using ng-include rather than by being routed to.  But, since you can have multiple `ng-outlet`s  (the new name for `ng-viewport`) on a page, there's no reason the shell and sidebar couldn't have been done using the router.  

So, I __recant__ my recommendation to use the default folder structure of the new router.  Instead, I recommend that we: 

1. Override the componentToTemplate function to remove the components folder as in the following.  In the case of the Hot Towel application, this alone will make everything route just fine.
2. Submit some code to the new angular router team which allows for overriding the default template path at a component-level.  This would then allow us to chunk together like capability like galleriesList, galleriesTree, galleryDetail, etc.  This could be a config parameter for each component like: `componentFolder = 'gallery/galleriesList'`, `componentFolder = 'gallery/galleriesTree'`, and `componentFolder = 'gallery/galleryDetail'`
3. Similar to bullet 2, submit some code to the new angular router team that allows you to specify at an App level what the root folder is for your components.  While bullet 1 is a quick fix for this, personally I'd rather not have to recommend that people override `$componentLoaderProvider`.  It's much easier to just have a config parameter like:  `rootComponentFolder = ''`

Here's an example of an overridden componentToTemplate that removes the components folder

    var componentToTemplate = function componentToTemplateDefault(name) {
        var dashName = dashCase(name);
        return './' + dashName + '/' + dashName + '.html';
    };

Phew!  Thanks for listening.

**Comment by johnpapa** [comment]
any more insight here? or is this something we close for now til there is an action?

**Comment by bleenders** [comment]
What I have is by no means an elegant solution.
For my dashboard app I wanted routes like /vendors, /vendors/create, /vendors/edit etc.
And store all the template files in a structure like:
- /components/vendors/vendors.html
- /components/vendors/vendors.js
- /components/vendors/create.html
- /components/vendors/create.js
- /components/vendors/edit.html
- /components/vendors/edit.js

I'm naming all the component as follows:

    $router.config([
            { path: '/', redirectTo: 'dashboard' },
            { path: 'dashboard', component: 'dashboard' },
            { path: 'vendors', component: 'vendors' },
            { path: 'vendors/create', component: 'vendorsCreate' },
            { path: 'vendors/create', component: 'vendorsEdit' },
    ]);

And for template mapping I'm using:

    $componentLoaderProvider.setTemplateMapping(function(name){
            var names = name.split(/(?=[A-Z])/);
            var folderStructure = '';
            for(var i = 0; i < names.length; i++){
               folderStructure += names[i];
                if( i < (names.length -1)){
                    folderStructure += '/';
                }
                if(names.length === 1){
                    folderStructure += '/' + names[i];
                }
            }
            folderStructure = folderStructure.toLowerCase();
            return '/components/'+ folderStructure + '.html';
        });

Controllers are named as follows:
- DashboardController
- VendorsController
- VendorsCreateController
- VendorsEditController

This allows me to group all route the files of a certain area of the application together and keep it somewhat modular.
It's probably not the best solution out there, but it works for me.

**Comment by robarchibald** [comment]
I'd say we should close this one and see what shakes out.  Several months ago the guidance from the Angular team was that the new Angular Router would be released with Angular 1.4 and would be the bridge between Angular 2 and Angular 1.  At the time, this seemed like an urgent issue to resolve, but since Angular 1.4 was released without the router, this is no longer urgent.  Now the router will first see the light of day in Angular 2.  From reading the angular team dev updates, it sounds like Angular 2 may be moving to a beta release sometime relatively soon, but I don't even want to speculate on what'll be in the router until at least the beta.  It has changed a TON since I last looked at it so there's no reason to believe that template path configuration won't be a part of the eventual release.  And, since Brad Green (I think) mentioned at Angular U that they are planning to look to the Angular Style Guide for guidance on how to structure apps, I hold high hope that the Angular 2 team will be open to ideas for where to place templates in folders rather than only in the components folder.

**Comment by zachlysobey** [comment]
I agree that this probably shouldn't get too much attention since the
details are still shaking out.

However, in response to: "Now the router will first see the light of day in
Angular 2", As I'm sure you know it *is* currently available. I'm using it
(in beta or whatever) for a 1.4 project, though somewhat regretting it as
it is poorly documented, buggy, and just not ready for production

The angular team meeting notes seem to imply that it *is *still being
worked on in 1.x.
From *https://docs.google.com/document/d/1xKEbydyUEOQ_gTbcbxy_k2myctG8EiVbeMgLgXTxIc0/edit
<https://docs.google.com/document/d/1xKEbydyUEOQ_gTbcbxy_k2myctG8EiVbeMgLgXTxIc0/edit>*
(non-router
bits removed):

August 3rd 2015
> Matias and Brian will examine router integration this week to get it to
> work with 1.4.x
> 27 July 2015
> Component Router update
> Landing Aux Routes this week
> Still working on 1.x build
> 20 July 2015
> Component Router in ng1 <https://github.com/angular/angular/pull/1707> -
> Shahar and Brian have been in discussions about this.
> Building the router
>
>    - Implementing Angular 1 bits
>
>



Zach Lysobey
(203) 984-4048

**Comment by robarchibald** [comment]
Fair enough Zach.  I guess its first release is slated for Angular 2 and
Angular 1.5, so whichever lands first will be where the first "official"
release of the new router is.  I actually tried using the new router back
in April I think.  I had everything all working and was actually pretty
happy with it.  I was planning to develop my new app using it until I saw
that they changed the name of  ng-viewport to ng-outlet, and made several
other changes that broke everything.  That was when I decided that waiting
was a good thing with the Angular Router.

Rob
